third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -690,22 +691,40 @@ impl<'a> FunctionGenerator<'a> {
         push_kind: Option<&AssignKind>,
     ) {
         let fun_ctx = ctx.fun_ctx;
-        // Compute the maximal prefix of `temps` which are already on the stack.
+        // Compute the maximal prefix of `temps` which are already on the stack. Also
+        // compute the offset from which on we need to flush the stack. See comments
+        // below for the need for flushing.
         let temps = temps.as_ref();
         let mut temps_to_push = temps;
+        let mut stack_to_flush = self.stack.len();
         for i in 0..temps.len() {
             let end = temps.len() - i;
             if end > self.stack.len() || end == 0 {
                 continue;
             }
             if self.stack.ends_with(&temps[0..end]) {
+                // We found 0..end temps which are already on top of the stack. The remaining ones
+                // need to be pushed.
                 temps_to_push = &temps[end..temps.len()];
+                // However, if any of the temps we found on the stack are used again after this program
+                // point we can't really consume them from the stack but need to flush them to registers.
+                for temp in &temps[0..end] {
+                    if ctx.is_alive_after(*temp) {
+                        stack_to_flush = std::cmp::min(
+                            stack_to_flush,
+                            self.stack
+                                .iter()
+                                .position(|t| t == temp)
```

Comment:
This is a bit inefficient, and is not clear enough to compensate.

May be better to start by first seeing how deep you need to flush to clear the stack of unlive temps, before looking for prefixes of `temps` that can be found in the remaining stack (the line 700 loop).

Btw, line 700-704 can be simplified to
```
for end in (1..=temps.len()).rev() {
```

Reply:
Done in #11568